tachyon/crypto/sumcheck/multilinear/data_structures.h

```diff
@@ -0,0 +1,173 @@
+ public:
+  using F = typename math::MultilinearExtension<Evaluations>::Field;
+
+  struct Term {
+    F coefficient;
+    std::vector<size_t> indexes;
+  };
+
+  ListOfProductsOfPolynomials() = default;
+  explicit ListOfProductsOfPolynomials(size_t num_variables)
+      : num_variables_(num_variables) {}
+  ListOfProductsOfPolynomials(ListOfProductsOfPolynomials& other) = default;
+  ListOfProductsOfPolynomials& operator=(ListOfProductsOfPolynomials& other) =
+      default;
+
+  // Adds a Term to the list
+  void AddTerm(
+      F& coefficient,
+      std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
+          evaluations) {
+    CHECK(!evaluations.empty());
+    size_t num_evaluations = evaluations.size();
+    max_evaluations_ = std::max(max_evaluations_, num_evaluations);
+    std::vector<size_t> indexed_evaluations;
+    indexed_evaluations.reserve(num_evaluations);
+    for (std::shared_ptr<math::MultilinearExtension<Evaluations>> evaluation :
+         evaluations) {
+      CHECK_EQ(evaluation->Degree(), num_variables_);
+      const math::MultilinearExtension<Evaluations>* evaluation_ptr =
+          evaluation.get();
+      auto it = raw_pointers_lookup_table_.find(evaluation_ptr);
+      if (it != raw_pointers_lookup_table_.end()) {
+        indexed_evaluations.push_back(
+            raw_pointers_lookup_table_[evaluation_ptr]);
+      } else {
+        size_t curr_index = flattened_ml_evaluations_.size();
+        flattened_ml_evaluations_.push_back(evaluation);
+        raw_pointers_lookup_table_[evaluation_ptr] = curr_index;
+        indexed_evaluations.push_back(curr_index);
+      }
+    }
+    terms_.push_back(Term{coefficient, indexed_evaluations});
+  }
+
+  // Evaluates all terms together on a an evaluation point for each variable:
+  // $$\sum_{i=0}^{n}coefficient_i\cdot\prod_{j=0}^{m_i}evaluation_{ij}$$
+  F Evaluate(std::vector<F>& point) const {
+    CHECK_EQ(point.size(), num_variables_);
+    return std::accumulate(
+        terms_.begin(), terms_.end(), F::Zero(),
+        [this, &point](F& sum, const Term& term) {
+          return sum +=
+                 term.coefficient *
+                 std::accumulate(
+                     term.indexes.begin(), term.indexes.end(), F::One(),
+                     [this, &point](F& product, size_t index) {
+                       return product *=
+                              this->flattened_ml_evaluations_[index]->Evaluate(
+                                  point);
+                     });
+        });
+  }
+
+  // Extract the max number of evaluations and number of variables
+  PolynomialInfo Info() const { return {max_evaluations_, num_variables_}; }
+
+ private:
+  // max number of evaluations for each term
+  size_t max_evaluations_ = 0;
+  // number of variables of the polynomial
+  // Note(ashjeong): this is equivalent to `kMaxDegree` of multilinear
+  // extensions
+  size_t num_variables_ = 0;
```

Comment:
If so, why do we have this member?

Reply:
It's needed to check line 75: `CHECK_EQ(evaluation->Degree(), num_variables_);` (note evaluation->Degree() gives us kMaxDegree for the given evaluation) and ensure that the number of variables in each individual evaluation given is valid. I put that note, since when I look at the MultilinearExtension or DenseMultilinearEvaluation files I was confused because the equivalent value of the intuitively labelled variable `num_variable` was instead non-intuitively named `kMaxDegree`.

Reply:
Q1. Why does `evaluation->Degree()` return `kMaxDegree`? Shouldn't it return the degree, not the maximum degree?
Q2. Can you explain why `kMaxDegree` is named in a non-intuitive manner?